gigantum/gigantum-client
# Fix custom base registry bug, cleanup base repos on error, logout on auth error

**Describe changes introduced by this PR (Bugs fixed, features added, docs updated, etc.)**:
This PR fixes a handful of small issues:

-  Fixed bug that prevented the use of non-Dockerhub registries for base images in custom base repositories. Now if it registry is listed in the base metadata file, it is properly rendered into the dockerfile.
- When an error occurs during base repository indexing, the repos are now removed and re-cloned on restart. This helps if there was an issue with the repo when it was cloned so a fetch will continue to fail.
-  When token validation fails, cached JWKs are removed and refetched. This helps in cases where a team server was restored OR when we push the token service update to gigantum hub prod and the jwks changes.
- Updated conda tests due to nltk version bump to get tests to pass.

**Please check if the PR fulfills these requirements**

  - PR is set to merge to integration
  - PR title is descriptive, as this becomes the commit message on squash and merge.
  - Any related issues are added to the extended commit message during squash and merge.
  - Tests have been added (for bug fixes/features)
  - Docs have been added/updated (for bug fixes/features)

**QA Test Steps and Exit Criteria**:
- a unit test was added for the non-dockerhub registry. This this is simply an issue with dockerfile rendering that probably is sufficient.
- To test base env cleanup, use the `config.yaml` file to point to a non-existing branch in our base images repo. This should result in the `~/gigantum/.labmanager/environment_repositories` directory getting deleted. Switch back to normal, restart, and you should be functioning and able to create a project.
- To test the jwks cleanup
  - Login
  - Using dev tools, remove the id and access token from browser storage
  - Refresh (should be at the server selection screen)
  - Open the cached JWKs and "corrupt" it by changing the `n` value or the `id`.
  - Log in -> you should see the contents of `~/gigantum/.labmanager/identity` get cleared out and then you log in successfully.
